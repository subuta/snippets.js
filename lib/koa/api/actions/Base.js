import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import {
  CONTROLLER_ACTIONS
} from 'lib/_utils/getControllerActions'

export default (props, children = []) => {
  let {
    model,
    routeConfig,
    path = '/',
    method = 'get',
    action
  } = props

  const Model = _.upperFirst(pluralize.singular(model))
  model = _.toLower(pluralize.singular(model))

  let middlewares = []

  const needsAuth = (() => {
    if (_.get(routeConfig, 'skipAuth') === true) {
      return []
    }
    return _.difference(CONTROLLER_ACTIONS, _.get(routeConfig, 'skipAuth', []))
  })()

  if (_.includes(needsAuth, action)) {
    middlewares.push('auth')
  }

  middlewares = _.isEmpty(middlewares) ? '' : middlewares.join(',') + ','

  return build`
    ${model}.${method}('${path}', ${middlewares} async (ctx) => {
      const {${Model}} = ctx.state.models
      ${children}
    })
  ` + s.EOL // append EOL to each line.
}
