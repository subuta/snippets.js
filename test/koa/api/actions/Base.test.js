/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'

import Base from 'lib/koa/api/actions/Base'

describe('koa/api/actions/Base', () => {
  it('should create koa action', () => {
    expect(format(Base({
      model: 'user',
      routeConfig: {},
      path: '/',
      method: 'post'
    }))).toMatchSnapshot()
  })
})
