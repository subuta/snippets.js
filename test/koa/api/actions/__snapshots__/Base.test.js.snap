// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`koa/api/actions/Base should create koa action 1`] = `
"user.post('/', async (ctx) => {
  const {User} = ctx.state.models
})
"
`;
